eZ/Bundle/EzPublishCoreBundle/EventListener/ContentDownloadRouteReferenceListener.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Bundle\EzPublishCoreBundle\EventListener;
+
+use eZ\Publish\API\Repository\Values\Content\Content;
+use eZ\Publish\API\Repository\Values\Content\Field;
+use eZ\Publish\Core\MVC\Symfony\Event\RouteReferenceGenerationEvent;
+use eZ\Publish\Core\MVC\Symfony\MVCEvents;
+use InvalidArgumentException;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\OptionsResolver\Options;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class ContentDownloadRouteReferenceListener implements EventSubscriberInterface
+{
+    const ROUTE_NAME = 'ez_content_download';
+    const OPT_FIELD_IDENTIFIER = 'fieldIdentifier';
+    const OPT_CONTENT = 'content';
+    const OPT_CONTENT_ID = 'contentId';
+    const OPT_DOWNLOAD_NAME = 'downloadName';
+    const OPT_LANGUAGE = 'language';
+    const OPT_VERSION = 'version';
+
+    public static function getSubscribedEvents()
+    {
+        return array(
+            MVCEvents::ROUTE_REFERENCE_GENERATION => 'onRouteReferenceGeneration'
+        );
+    }
+
+    /**
+     * @throws \InvalidArgumentException If the required arguments are not correct
+     */
+    public function onRouteReferenceGeneration( RouteReferenceGenerationEvent $event )
+    {
+        $routeReference = $event->getRouteReference();
+
+        if ( $routeReference->getRoute() != self::ROUTE_NAME )
+        {
+            return;
+        }
+
+        $options = $routeReference->getParams();
+
+        $resolver = new OptionsResolver();
+        $resolver->setRequired( [ self::OPT_CONTENT, self::OPT_FIELD_IDENTIFIER ] );
+        $resolver->setDefined( [ self::OPT_VERSION, self::OPT_LANGUAGE, self::OPT_DOWNLOAD_NAME, self::OPT_CONTENT_ID ] );
+        $resolver->setAllowedTypes( self::OPT_CONTENT, 'eZ\Publish\API\Repository\Values\Content\Content' );
+        $resolver->setAllowedTypes( self::OPT_FIELD_IDENTIFIER, 'string' );
+        $resolver->setDefault(
+            self::OPT_CONTENT_ID,
+            function( Options $options ) {
+                return $options[self::OPT_CONTENT]->id;
+            }
+        );
+
+        $options = $resolver->resolve( $options );
+
+        $field = $options[self::OPT_CONTENT]->getField( $options[self::OPT_FIELD_IDENTIFIER] );
+        if ( !$field instanceof Field )
+        {
+            throw new InvalidArgumentException(
+                sprintf(
+                    "The '%s' parameter did not match a known Field",
+                    self::OPT_FIELD_IDENTIFIER
+                )
+            );
+        }
+
+        // use the value's filename if not given
+        if ( !isset( $options[self::OPT_DOWNLOAD_NAME] ) )
+        {
+            $options[self::OPT_DOWNLOAD_NAME] = $field->value->fileName;
+        }
```

Comment:
I can't do that until I have the field, and I get the field from the options *after resolution*.

I can of course test the parameters manually, but it would kind of duplicate what is done by the OptionsResolver itself, would it not ?